feat: swap OutlineNav into sidecar

zchsh opened the pull request:
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Adds sidecar scrolling to `SidebarSidecarLayout`, and starts using the new `OutlineNav` component.

## 🤷 Why

To implement new layout behaviour for our sidecar slot, and to use the newly designed and implemented `OutlineNav` component.

## 🛠️ How

- Implements `SidecarScrollContainer` for use in `SidebarSidecarLayout`
    - This component handles the scrollable container logic for `sidecarSlot` contents
- Updates `SidebarSidecarLayout` to use `SidecarScrollContainer`, and tweaks styles accordingly
- Swaps in `OutlineNavFromHeadings` for `TableOfContents` as sidecar contents
    - Most uses are covered by making the swap in `SidebarSidecarWithToc`
    - Note: the Integrations component view did not use `SidebarSidecarWithToc`, so it's one case we have to update outside of `SidebarSidecarWithToc`. For now we swap in `OutlineNavFromHeadings`, looking forward we'll be swapping in new `OutlineNav` logic in #1667. 

## 📸 Design Screenshots

<!--
Include a screenshot or two and a link to the designs you referenced with this code. These are both helpful context for reviewers to understand what designs you were looking at when you were putting this code together.
-->

### Design

<img width="1270" alt="CleanShot 2023-02-23 at 14 48 20@2x" src="https://user-images.githubusercontent.com/4624598/221014364-e6e77e59-8cba-4138-a60c-08ed46dbb7a5.png">

### Implementation

<img width="1920" alt="implementation" src="https://user-images.githubusercontent.com/4624598/221016878-152bc849-6f18-4ac0-a5f1-1e0eabc83218.png">

## 🧪 Testing

Pages that previously used `TableOfContents` should all be updated to use `OutlineNav`. Active section highlighting should function as it did previously.

Specifically, the following pages should now be rendering `OutlineNav`:

- Product landing pages, such as [/waypoint][/waypoint]
- Product tutorials landing pages, such as [/waypoint/tutorials][/waypoint/tutorials]
- All docs document pages, such as [/vault/api-docs/auth/aws][/vault/api-docs/auth/aws]
    - ⭐ This page in particular shows overflow behaviour quite well, as there are tons of page headings!
- All tutorials document pages, such as [/waypoint/tutorials/get-started-docker/get-started-docker][/waypoint/tutorials/get-started-docker/get-started-docker]
- Onboarding tutorial pages, such as [/onboarding/hcp-vault-week-2/static-secrets][/onboarding/hcp-vault-week-2/static-secrets]
- WAF tutorial pages, such as [/well-architected-framework/com/cloud-operating-model][/well-architected-framework/com/cloud-operating-model]
- Integration component pages, such as [/waypoint/integrations/hashicorp/aws-ecs/latest/components/platform][/waypoint/integrations/hashicorp/aws-ecs/latest/components/platform]
- [/well-architected-framework][/well-architected-framework]

## 💭 Anything else?

Not at the moment!

[task]: https://app.asana.com/0/1203810296555970/1203964362474949/f
[preview]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/
[/waypoint]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/waypoint
[/waypoint/tutorials]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/waypoint/tutorials
[/vault/api-docs/auth/aws]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/vault/api-docs/auth/aws
[/waypoint/tutorials/get-started-docker/get-started-docker]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-docker
[/onboarding/hcp-vault-week-2/static-secrets]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/onboarding/hcp-vault-week-2/static-secrets
[/well-architected-framework/com/cloud-operating-model]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/well-architected-framework/com/cloud-operating-model
[/waypoint/integrations/hashicorp/aws-ecs/latest/components/platform]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/waypoint/integrations/hashicorp/aws-ecs/latest/components/platform
[/boundary/api-docs]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/boundary/api-docs
[/hcp/api-docs/packer]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/hcp/api-docs/packer
[/well-architected-framework]: https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/well-architected-framework

braydenlove commented:
This is great @zchsh , a few questions and VQA items for you.
- Can we remove the gradient once the user scrolls to the point where they can see the overflow items? 
<img width="968" alt="Showing gradient even when scrolled to the bottom" src="https://user-images.githubusercontent.com/81585949/220450371-f98dc003-86c4-4085-8dfe-f3684e2260f3.png">

- I am sure I am missing something so keep me informed :). I am not sure what the empty space is below the last sidecar item. Are we able to remove that and simply end the scrollable area at the bottom (don't extend it to the bottom of the page)?
<img width="1920" alt="red rectangle shows empty space in question" src="https://user-images.githubusercontent.com/81585949/220450588-d0090352-ee9d-492d-aba4-1775b2be58b6.png">

- When the viewport is on the smaller side we get into a situation where the scrollbar for the sidecar is sitting on the edge of the page (you nailed it based on my designs). In practice that feels a bit awkward when actually scrolling, curious if we can give it a bit of breathing room 16px margin to start)?
<img width="1351" alt="scrollbar sitting on right side of page" src="https://user-images.githubusercontent.com/81585949/220450694-914205b6-1e81-481b-8ca5-ec1b56933aba.png">

- We shouldn't go to two lines on wrapping (instead we should truncate with a ...). Here is an [example](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/terraform/registry/api-docs).

Let me know if it is easier to jump on a zoom to discuss. Also if you need specific Figma examples hit me up! CC @kellyschummer

zchsh commented:
@braydenlove Thanks for the feedback, these are great! 💖 

- re: Removing gradient once at bottom - yep, I should be able to add this logic in! Will update once I have it done.
- re: Blank space at bottom - this was kind of an MVP version of avoiding the bottom contents of the nav being shrouded without having to add scroll position logic to remove the gradient. Once we add the logic to remove the gradient when scrolled to the bottom, it'll no longer be necessary 👍
- re: Scroll bar spacing - good call, i'll add 16px padding the sidecar scroll bar and the right edge of the viewport
- re: Truncation - ah, I missed this, thought we were getting rid of truncation - will add in CSS-based truncation so items are always at most one full line

zchsh commented:
@braydenlove I have some updates here, hopefully covering all the feedback items so far - let me know what you think 🙇

### Gradients when user scrolls

I've updated so the gradient scrim at the bottom disappears when the sidecar is fully scrolled. I've also removed the bottom padding that was there previously, so the bottom of the sidecar contents are always at the bottom of the sidecar container.

As well, I tried out adding a gradient scrim at the top of the container, with similar logic for dynamically showing & hiding it based on scroll position. Let me know what you think, easy to remove that if we'd prefer to only have the scrim at the bottom.

Here's a quick demo:

https://user-images.githubusercontent.com/4624598/220762883-3a0b1c6e-1689-4ae7-a55e-9f0c06327e96.mp4

### Scrollbar position on smaller viewports

I've added `16px` padding as requested. Looks good to me! 🙌 

![2023-02-22-padding-added](https://user-images.githubusercontent.com/4624598/220763013-c9e6b985-584d-49ff-aa0f-d7a797f3d9c1.png)

### Truncation

I've updated things so items are truncated. Here's [the example you mentioned](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/terraform/registry/api-docs):

![2023-02-22-truncation](https://user-images.githubusercontent.com/4624598/220763178-f4cd2eb6-928b-404f-9854-f9ce2dd11a54.png)

alexcarpenter commented:
I have yet to ever use this method, but there supposedly is this native CSS way for showing and hiding gradients on scroll that I thought was worth sharing https://css-tricks.com/books/greatest-css-tricks/scroll-shadows/

Again, not sure if it works 100% for our use case here.

braydenlove commented:
@zchsh really awesome work, you hit everything I asked for and added that top gradient which is so nice <chef's kiss>. That all looks good to me.

One thing I noticed was there is some sticky issues when scrolling to the bottom of a page (on sidecars that don't exceed the height of the page). When you scroll to the bottom the sidecar scrolls out of view. Curious if we are able to lock it in place in a scenario like [this](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/terraform/cloud-docs/architectural-details/data-security#blob-storage-encryption)?

Ping Kelly or Mike while I am away if you are looking to merge this asap. If not I will follow up Monday AM as I know you are off for the day. Thanks, awesome work!

braydenlove commented:
One last thing I noticed... it looks like you aren't able to click on the top sidecar item (page title). Here is a [link](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/packer/tutorials/hcp-get-started/hcp-immediately-revoke-image-iterations#next-steps).

zchsh commented:
Oops - re: non-clickable top item, i had some CSS i'd written locally but forgot to push up 😅 Should be working now! I'll look into the shorter sidecar sticky issue - that seems like something we should be able to address, will follow up here...

zchsh commented:
@kendallstrautman @kellyschummer This PR swaps in the recently implemented `OutlineNav` (from #1662) for all existing instances of `TableOfContents`. Let me know what ya'll think 🙇

kellyschummer reviewed (approved):
🥇 to these updates !! @zchsh Thank you!

zchsh commented:
@im2nguyen Tagging you here for content review - we're looking for someone on the content side to click around and ensure that this revised sidecar works as expected, and everything is still feeling good. There's a selection of preview links in the PR description. Let me know what you think 🙇

zchsh commented:
Minor note on changes since requesting review: just had to tweak a `margin-top` value to get the layout to work as expected when alert bars are present. Quick demo:

https://user-images.githubusercontent.com/4624598/221030123-4da5b4bf-2593-4b4e-85ea-973069108999.mp4

zchsh commented:
Ack, in all the focus on design review I've forgotten to look into [the scroll shadows approach Alex mentioned](https://css-tricks.com/books/greatest-css-tricks/scroll-shadows/) (thanks for that @alexcarpenter! 💖 ). Going to look into that tomorrow AM, might open a separate PR to keep the ball rolling on this one 🎢

ashleemboyer reviewed (commented):
Just a few comments! And also something I noticed while testing -- when I navigate from a page with a long sidecar + gradient to a page with a short sidecar, the gradient is still present on the short sidecar. I started at [/vault/api-docs/auth/aws](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/vault/api-docs/auth/aws), then navigated to [/vault/api-docs/secret/key-management/gcpkms](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/vault/api-docs/secret/key-management/gcpkms) using the sidebar.

zchsh commented on src/layouts/sidebar-sidecar/sidebar-sidecar-layout.module.css:
```diff
@@ -122,16 +126,27 @@ for further details.
+	--sidecar-position-top: calc(
+		var(--sticky-bars-height) + var(--main-area-padding-top) +
+			var(--breadcrumb-bar-height)
+	);
```
@ashleemboyer Great call 💖 I [considered it](https://github.com/hashicorp/dev-portal/pull/1661/commits/e2dad01ecde1e8c505d244b9d72576d12032f7c0#diff-0c99b3fcf37ed6bb4819102c69efa12cc6c55e69153cd52fede833b259810dc1R141) but was unsure about the tradeoff of trying to dive in and refactor the layout a bit, since the downside here would be a bit of misalignment rather than something totally broken, and it seemed _possible_ the refactor could spiral into deeper work.

That being said, deadline here isn't tight, so I took another look and have just pushed up some refactoring to `sidebar-sidecar.module.css`. Here's a quick summary, `before` on the left, `after` on the right:

<img width="2560" alt="before-after" src="https://user-images.githubusercontent.com/4624598/221188740-4c556c33-5e6b-406c-82e5-4a27b438b02a.png">

I think this should work as expected, and continue to match designs, while removing the brittleness of relying on a specific pixel value for the breadcrumb bar height. This should also help prevent breadcrumb wrapping in some cases:

| Before | After |
| - | - |
| ![before](https://user-images.githubusercontent.com/4624598/221191226-38388955-8926-41d1-8f99-1fe77d3ae37e.png) | ![after](https://user-images.githubusercontent.com/4624598/221191231-49ebfa3d-fe7c-4c43-a5c0-ac0512222297.png) |

Let me know what you think 🙇

zchsh commented on src/layouts/sidebar-sidecar/components/sidecar-scroll-container/sidecar-scroll-container.module.css:
```diff
@@ -0,0 +1,80 @@
+.root {
+	--gradient-scrim-height: 58px;
+	--gradient-scrim-transition: opacity 0.25s;
```
absolutely, great catch! ✅ updated to wrap the `transition` declarations.

zchsh commented on src/layouts/sidebar-sidecar/components/sidecar-scroll-container/sidecar-scroll-container.module.css:
```diff
@@ -0,0 +1,80 @@
+	padding-left: 1px;
+	padding-right: 1px;
+	padding-top: 1px;
+	padding-bottom: 1px;
```
for sure - ✅ updated

zchsh commented on src/layouts/sidebar-sidecar/sidebar-sidecar-layout.module.css:
```diff
@@ -113,6 +113,10 @@ for further details.
 	padding-bottom: 128px;
+
+	@media (--dev-dot-sidecar-up) {
+		padding-right: 16px;
```
Great call - with the work I did as part of figuring out how to decouple parts of the layout from the breadcrumb bar height, it turned out `--main-area-padding-right` and `--main-area-padding-left` were no longer necessary to share between components (aka they were single-use CSS properties, nothing else was using them). So I've removed the CSS properties, since that felt like it cleaned things up a bit 👍

zchsh commented on src/layouts/sidebar-sidecar/sidebar-sidecar-layout.module.css:
```diff
@@ -52,12 +52,12 @@ for further details.
 	&.mainWidth-narrow {
 		--main-element-max-width: 680px;
-		--sidecar-width: 288px;
+		--sidecar-width: 312px;
```
Good call, updated! ✅

zchsh commented on src/layouts/sidebar-sidecar-with-toc/index.tsx:
```diff
@@ -6,17 +6,16 @@
+import { OutlineNavFromHeadings } from 'components/outline-nav/components'
 
 export type SidebarSidecarWithTocProps = SidebarSidecarLayoutProps & {
 	headings: TableOfContentsHeading[]
```
I don't think so - but I think it's something we'd like to do eventually. It would involve refactoring `getStaticProps` for every page that uses this layout to change what's being returned from the server (currently `TableOfContentsHeading`, we probably want it to be `OutlineHeading`). This felt like more of a DX change than anything else, and the effort involved seems non-trivial, so I have a separate follow-up task 🎟 [Clean up headings-related OutlineNav uses, to move formatting closer to server](https://app.asana.com/0/1203810296555970/1204024966433303/f) to tackle it.

Let me know if that's helpful! 🙇

zchsh commented:
Thanks for the awesome feedback @ashleemboyer ! 💖 Have done my best to address each item. Here's a summary of the blocking items:

- 🚧 Breadcrumb bar height - addressed [here](https://github.com/hashicorp/dev-portal/pull/1661#discussion_r1117010425)
- 🚧 `prefers-reduced-motion` - have wrapped the `transition` in this, thanks for catching this 🙌

As well, re: the gradient scrim inaccuracy on route changes from a long to short sidecar, I've addressed this by checking whether the gradient scrim is needed when `router.asPath` changes. So now when client-side navigating from a page with a long sidecar to a page with a short sidecar, the gradient scrim should disappear as expected.

Incidentally, cc @alexcarpenter, I briefly tried the CSS-only approach to the scroll container gradient scrims. The block I ran into was that it seems to only work for gradient _backgrounds_. In this case, we want the gradient to layer above the content of the container. I couldn't think of a way to do that with a CSS-only approach.

ashleemboyer commented:
> Thanks for the awesome feedback @ashleemboyer ! 💖 Have done my best to address each item. Here's a summary of the blocking items:
> 
> * 🚧 Breadcrumb bar height - addressed [here](https://github.com/hashicorp/dev-portal/pull/1661#discussion_r1117010425)
> * 🚧 `prefers-reduced-motion` - have wrapped the `transition` in this, thanks for catching this 🙌
> 
> As well, re: the gradient scrim inaccuracy on route changes from a long to short sidecar, I've addressed this by checking whether the gradient scrim is needed when `router.asPath` changes. So now when client-side navigating from a page with a long sidecar to a page with a short sidecar, the gradient scrim should disappear as expected.
> 
> Incidentally, cc @alexcarpenter, I briefly tried the CSS-only approach to the scroll container gradient scrims. The block I ran into was that it seems to only work for gradient _backgrounds_. In this case, we want the gradient to layer above the content of the container. I couldn't think of a way to do that with a CSS-only approach.

@zchsh re-reviewing now! 😊

ashleemboyer commented:
@zchsh still reviewing the code, but wanted to leave this here before finishing that in case you're able to debug between now and then!

I'm no longer seeing the gradient at the bottom of the sidecar when I land on the [/vault/api-docs/auth/aws](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/vault/api-docs/auth/aws). Might be because of the layout arranging you did? Seems like it just needs a lil' height adjustment is all! My viewport is `1792px` wide by `1009px` tall in this screenshot.

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/43934258/221227695-b7541a5d-c37e-4c54-9b8b-89692bca57c2.png">

zchsh commented:
@ashleemboyer Ah yep - that's a direct result of the layout changes, and is intentional! You'd see a similar effect if for example there was a global alert banner at the top of the page. Once you start scrolling, the sidecar comes fully into view.

The only way i can think to prevent this is to have the sidecar height based on the breadcrumb bar height - but as you noted I think that's something we'd want to avoid 👍

ashleemboyer commented:
> @ashleemboyer Ah yep - that's a direct result of the layout changes, and is intentional! You'd see a similar effect if for example there was a global alert banner at the top of the page. Once you start scrolling, the sidecar comes fully into view.
> 
> The only way i can think to prevent this is to have the sidecar height based on the breadcrumb bar height - but as you noted I think that's something we'd want to avoid 👍

@zchsh Ah, got it! Seems minor, so maybe we're good on this for now.

---

Just came across another layout-y issue between `1280px` and about `1425px` wide, where content is getting cut off on the left & right side. Seems also related to the rearrangement?

![CleanShot 2023-02-24 at 11 12 14@2x](https://user-images.githubusercontent.com/43934258/221230080-dc86c4b5-2a62-4181-b6f0-686fa95502a2.png)

ashleemboyer commented:
@zchsh sorry for the multiple rapid fire comments 😅 

I found an example of a _**really**_ long breadcrumb under HCP docs: [/hcp/docs/hcp/supported-env/azure](https://dev-portal-git-zsnew-outline-in-sidecar-hashicorp.vercel.app/hcp/docs/hcp/supported-env/azure)

<img width="1325" alt="CleanShot 2023-02-24 at 11 22 36@2x" src="https://user-images.githubusercontent.com/43934258/221231771-921ef31e-0e17-4857-b868-78efaa37f1dc.png">

I think this is something we'd need design approval for, and additionally probably more time to work on _when_ `BreadcrumbBar` truncates. I'm OK with removing my blocking comment about the hardcoded sidecar placement for now! Maybe we could add a backlog ticket for continued exploration with more stable placement?

zchsh commented:
@ashleemboyer Great point, the layout changes do affect the breadcrumb width. One option here, that would reflect the previous positioning if we want it to, could be to change the `max-width` limit for the breadcrumb group. Here's a visual summary:

| Current | Limited width breadcrumb |  Full width breadcrumb |
| - | - | - |
| ![current](https://user-images.githubusercontent.com/4624598/221235617-4ed587a2-2b8a-4a46-b5b8-2a6088b56e9a.png) | ![limited-width](https://user-images.githubusercontent.com/4624598/221235616-a049dd7e-61b5-4f16-b9f3-a8a36b940e33.png) | ![full-width](https://user-images.githubusercontent.com/4624598/221235611-aae4eeee-96a2-427c-8e19-33f652404299.png)

@kellyschummer  - looping you in here and re-requesting review. I wasn't sure from designs we have - do you think we'd want to go with the "full width" or "limited width" breadcrumb as above?

kellyschummer commented:
I'd like to move forward with the limited width breadcrumbs, as this isn't much of a departure from the current wrap and gives adequate breathing room around all elements as intended. I also understand that there's some breadcrumb updates on the horizon that will further improve this view.  Thank you @ashleemboyer @zchsh for catching and exploring 💯 ❤️ options.

zchsh commented:
Note: I've updated to use the "limited width" breadcrumb mode, but builds here are failing, I think for unrelated reasons, so previews aren't yet reflecting the "limited width" breadcrumb area.

braydenlove reviewed (approved):
I just VQA'ed a variety of templates on chrome, safari, Firefox and Edge (PC). All the bugs we (_design_) found look solved for and this is feeling great. Thanks for the hard work, if we notice more things in the future we can always create follow up tickets. As far as I can see this morning this is working very well!

heitzke commented:
Agreed here, thanks for all the work to get this together @zchsh and it's looking good

ashleemboyer reviewed (approved):
LGTM! 👍🏻 ✨

kendallstrautman commented on src/layouts/sidebar-sidecar/components/sidecar-scroll-container/sidecar-scroll-container.module.css:
```diff
@@ -0,0 +1,83 @@
+	--gradient-scrim-height: 58px;
+	--gradient-scrim-transition: opacity 0.25s;
+
+	/* TODO: we need a --token-color-surface-primary-transparent, or something! */
```
💡 Good idea

im2nguyen reviewed (approved):
Ashlee and Brayden did an awesome job with VQA. This looks great! 🚀
